Automattic/jetpack
# Add additional properties to WC Analytics events

~Please do not merge this PR yet. Changes in this PR depend on the newly added action in WC Pay. We need to merge these changes after releasing the WC Pay version 2.9.~ WC Pay version 2.9 has been [released](https://github.com/Automattic/woocommerce-payments/releases/tag/2.9.0) and I've updated this PR to work with previous versions of WC Payments in 4e9a4fb071145687942c8bdb6a552c6026e483ff

Fixes https://github.com/woocommerce/woocommerce-admin/issues/7301

## Changes proposed in this Pull Request:

- Supports `array` value to be added as an event property https://github.com/Automattic/jetpack/commit/4f2d1bce8ded004302e0bd78a4340e233df3c440 - This is needed to add `payment_options` value.
-  Added additional properties to `woocommerceanalytics_product_purchase` and `woocommerceanalytics_product_checkout_view`. The additional properties are listed in the [issue](https://github.com/woocommerce/woocommerce-admin/issues/7301).
- One notable change is that it uses an [action](https://github.com/Automattic/woocommerce-payments/pull/2610).
 from WC Pay when the express checkout methods are enabled.

## Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



We're adding the following properties to `woocommerceanalytics_product_purchase` and `woocommerceanalytics_product_checkout_view` events.

#### woocommerceanalytics_product_purchase

**payment_option**: the option that was used. Eg Stripe or COD
**express_checkout**: if the express checkout option was used: Eg: Apple Pay or Null
**guest_checkout**: if the guest checkout was used
**create_account**: check if the shopper created an account (Y/N)
**device**: mobile/desktop

#### woocommerceanalytics_product_checkout_view

**payment_options**: an array with the options available. Eg [paypal, stripe, cod, amazon]
**express_checkout**: an array with the options available. Eg [apple_pay] or [null] if they don’t display them
**guest_checkout**: Yes/No. Identify if the store allows users to checkout without an account
**create_account**: Yes/No Identify if the store allows users to create an account on checkout
**device**: mobile/desktop

## Testing instructions:

Please use [WC Pay 2.9.0](https://github.com/Automattic/woocommerce-payments/releases/tag/2.9.0)

If you're new to WC Analytic testing, please perform the following steps to enable WC Analytic.

1. Install and activate WooCommerce.
2. Finish onboarding wizard and enable `Direct bank transfer` in WooCommerce -> Settings -> Payments.
2. Install and activate Jetpack.
3. Connect your site to Jetpack
4. WC Analytic doesn't track in test/dev mode. Open this [file](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php#L41) and add `return true` to force WC Analytic to track during the testing.
5. Click on a product on `/shop` to see the details.
6. Open your browser inspector -> Network -> search for "t.gif" then refresh.
7. You should see `t.gif` request.

If you don't see `t.gif` request for some reason, make sure WooCommerce Analytics module is activated in `/wp-admin/admin.php?page=jetpack_modules`


**Case 1** -- Without WC Payment installed/activated

1. Make sure WC Payment plugin is **not** installed/activated.
2. Add a product to the cart.
3. Open browser inspector -> Network and search for `t.gif`
4. Continue to the checkout page.
5. You should see `t.gif` request.
6. Click on the request and confirm that the request URL contains all the required properties for `woocommerceanalytics_product_checkout_view` event.
7. Click [ Place order ] button with `Direct bank transfer` selected.
8. You should see another `t.gif` request on the `Order received` page for `woocommerceanalytics_product_purchase` event. Confirm that the required properties are included in the URL.

Bonus: Add multiple products and confirm you get `t.gif` request for each product.

![Screen Shot 2021-08-25 at 11 46 03 AMth_](https://user-images.githubusercontent.com/4723145/130847883-48bd2254-ca93-4e66-b426-962ee94f15b7.jpg)

You can use tools like https://www.freeformatter.com/url-parser-query-string-splitter.html to parse the URL into a table.

**Case 2** -- WC Payment activated + express checkout methods enabled.

1. Use ngrok to get `https` and publicly accessible address.
2. Install, activate, and connect WC Payment
3. Make sure `Express Checkouts` are enabled in Payments -> Settings
4. Repeat steps 2~8 from Case 1.
5. Confirm both events have `express_checkout` value.

`woocommerceanalytics_product_checkout_view` should have either `google_pay` or `apple_pay`
`woocommerceanalytics_product_purchase` should have `null`

**Case 3** -- WC Payment activated and express checkout methods disabled.

1. Use ngrok to get `https` and publicly accessible address.
2. Install, activate, and connect WC Payment
3. Navigate to Payments -> Settings and disable `Express Checkouts`
4. Repeat steps 2~8 from Case 1
5. Confirm `express_checkout` value is `null` in both events.

**Case 4** -- WC Payment activated + express checkout methods enabled + checkout using Google/Apple pay.

1. Use ngrok to get `https` and publicly accessible address.
2. Install, activate, and connect WC Payment
3. Navigate to Payments -> Settings and disable `Express Checkouts`
4. Create a new product and set the price to $0.5 (minimum price to use google pay).
5. Add the product to the cart and contitnue to the checkout page.
6. Click Google Pay or Apple Pay to checkout. 
7. Confirm `express_checkout` value.

both events should have `google_pay` or `apple_pay`

**Case 5** -- WC Payment version lower than 2.9.0.

The action used in this PR was added in WC Payment version 2.9.0. If the user has WC Payment version lower than 2.9.0, then the action should not be used.

1. Install & activated WC Payment version lower than 2.9.0
2. Add a product to the cart and continue to the checkout page.
3. On the checkout page, right-click on the page and click "View Page Source"
4. Scroll to the bottom of the page and confirm the event JS script is not using `wp.hooks.addAction`
